File: 프로그래머스/Level2/괄호 변환.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

bool correct(string s){
    stack<char> st;
    
    for(char i : s){
        if(i == '(') st.push(i);
        else if(i == ')'){
            if(st.empty()) return false;
            else st.pop();
        }
    }
    if(st.empty()) return true;
    else return false;
}

string recursion(string s){
    
    if(s == "") return s;
     
    string u = "";
    string v = "";
    int cnt1 = 0, cnt2 = 0, index = 0;
    string resultString= "";
    
    for(char i : s){
        if(i == '(') cnt1 ++;
        else cnt2 ++;
        index ++;
        
        if(cnt1 == cnt2) break;
    }
    
    u = s.substr(0,index);
    v = s.substr(index);
    
    if(correct(u)) resultString += u + recursion(v);
    else{
        resultString += '(';
        resultString += recursion(v);
        resultString += ')';
        
        u = u.substr(1,u.size()-2);
            
        for(int i: u){
            if(i == '(') resultString += ')';
            else resultString += '(';
        }
    }
    return resultString;                
}

string solution(string p) {
    string answer = "";
    
    answer = recursion(p);

    return answer;
}

File: 프로그래머스/Level2/괄호 변환.js
function solution(p) {
    if(p === "") return p; 
    let u = '', v = '', left = 0, right = 0;

    for(let i = 0; i<p.length; i++){
        p[i] === '(' ? left ++ : right ++;

        if(left === right){ 
            u = p.slice(0, i+1);
            v = p.slice(i+1, p.length); 

            if(checkRightString(u)) return u + solution(v);
            else return `(${solution(v)})${u.slice(1, u.length-1).split('').map(elem => elem === '(' ? ')' : '(').join('')}`;
        }
    }
}

function checkRightString(str){
    const stack = [];

    for(let i of str){
        if(i === '(') stack.push(i);
        else{
            if(stack.length <= 0) return false
            else stack.pop();
        }
    }

    return true;
}
